marlowe-playground-client/src/LoginPopup.purs

```diff
@@ -0,0 +1,103 @@
+module LoginPopup where
+
+import Prelude
+import Auth (AuthRole)
+import Control.Monad.Except (runExcept)
+import Data.Either (Either(..), hush)
+import Data.Maybe (Maybe(..))
+import Data.Traversable (for_)
+import Effect (Effect)
+import Effect.Aff (Aff, finally, makeAff, nonCanceler)
+import Effect.Class (liftEffect)
+import Effect.Ref (Ref)
+import Effect.Ref as Ref
+import Foreign.Generic.Class (encode, decode)
+import Web.Event.Event (EventType(..), Event)
+import Web.Event.EventTarget (EventListener, addEventListener, eventListener, removeEventListener)
+import Web.HTML as Web
+import Web.HTML.Location (replace)
+import Web.HTML.Window (outerHeight, outerWidth)
+import Web.HTML.Window as Window
+import Web.HTML.WindowExtra (close, postMessage)
+import Web.Socket.Event.MessageEvent as MessageEvent
+
+-- This function is intended to be called between a parent window. It creates a popup to the
+-- the Github auth page and waits for the popup to send a message with the current AuthRole
+openLoginPopup :: Aff AuthRole
+openLoginPopup = do
+  window <- liftEffect Web.window
+  let
+    popupHeight = 620
+
+    popupWidth = 600
+
+    githubLoginPage = "/api/oauth/github"
+
+    features :: Aff String
+    features =
+      liftEffect do
+        top <-
+          outerHeight window
+            <#> \windowHeight -> windowHeight / 2 - popupHeight / 2
+        left <-
+          outerWidth window
+            <#> \windowWidth -> windowWidth / 2 - popupWidth / 2
+        pure $ "width="
+          <> show popupWidth
+          <> ",height="
+          <> show popupHeight
+          <> ",top="
+          <> show top
+          <> ",left="
+          <> show left
+          <> ",menubar=no,status=no,location=no"
+
+    decodeMessageEvent :: Event -> Maybe AuthRole
+    decodeMessageEvent event = do
+      data' <- MessageEvent.data_ <$> MessageEvent.fromEvent event
+      hush <<< runExcept <<< decode $ data'
```

Comment:
just a minor thing but I feel like (especially given the name) it would be better to return an `Either` and ignore the error at the site of use rather than here.

Reply:
Mhmhm, for this PR I think it makes more sense leaving it as it currently is. I tried changing it, but fromEvent returns a Maybe, so I should use `note` to indicate failure reason, and that implies that I need to choose a new error type of the sum of fromEvent and decode or to `note` fromEvent with a `ForeignError`.